Userland/Libraries/LibWeb/CSS/StyleProperties.cpp

```diff
@@ -672,6 +672,17 @@ Optional<CSS::PointerEvents> StyleProperties::pointer_events() const
     return keyword_to_pointer_events(value->to_keyword());
 }
 
+Optional<Variant<CSS::Length, double>> StyleProperties::tab_size() const
+{
+    auto value = property(CSS::PropertyID::TabSize);
+    if (value->is_length())
+        return value->as_length().length();
+
+    if (value->is_number())
+        return value->as_number().number();
+    return {};
+}
```

Comment:
This is good, but we want to handle `calc()` too. Confusingly, `calc()` is [specified](https://www.w3.org/TR/css-values/#math) as going anywhere that's a numeric type - `<integer>`, `<length>`, `<percentage>` etc. And yes, this is a bit of a pain. :sweat_smile: 

Here, `<length>` or `<number>` could be `calc()`, so we need to handle either. That's something like:

```c++
if (value->is_math()) {
	auto& math = value->as_math();
	if (math.resolves_to_length())
		return math.resolve_length(layout_node);
	if (math.resolves_to_number())
		return math.resolve_number();
}
```

That also means passing in a `Layout::Node&` to the function so that you can use it.

However, the other way would be to store the `calc()` for later. Technically I think that's more correct - I always get a bit confused about when we're guaranteed to be able to fully resolve a `calc()` value. To do that, you'd change your `Variant<CSS::Length, double>` everywhere to a `Variant<LengthOrCalculated, NumberOrCalculated>`, and the code snippet above would be more like this, without needing a `Layout::Node`:

```c++
if (value->is_math()) {
	auto& math = value->as_math();
	if (math.resolves_to_length())
		return LengthOrCalculated { math };
	if (math.resolves_to_number())
		return NumberOrCalculated { math };
}
```

`FooOrCalculated` is basically a `Variant<Foo, CSSMathValue-that-resolves-to-Foo>`.

Reply:
Ah OK, thanks for the detailed explanation. This will help others understand the process if they ever check this PR. Still wrapping my head around all the cases I need to handle. I'll keep in mind that `calc()` could go anywhere for those numeric types for next time. Being a full-stack engineer I still forget about all these css shenanigans lol. 

I think method two sounds more compelling, so I may give that a shot. Thanks.